### PR TITLE
RavenDB-13163 creating documents with very long names

### PR DESCRIPTION
**Pull request by aviv86**
https://issues.hibernatingrhinos.com/issue/RavenDB-13163

### CODE REVIEW

**Review by grisha-kotler — CHANGES_REQUESTED** [review]
Need another test for attachments

File: src/Raven.Server/Documents/AttachmentsStorage.cs [hunk 1]
```diff
@@ -166,6 +166,11 @@ public IEnumerable<DynamicJsonValue> GetAttachmentsMetadataForDocumentWithCounts
                 if (TableValueToFlags((int)DocumentsTable.Flags, ref tvr).HasFlag(DocumentFlags.Artificial))
                     throw new InvalidOperationException($"Cannot put attachment {name} on artificial document '{documentId}'.");
 
+                if (documentId.Length + name.Length > DocumentIdWorker.MaxIdSize)
```
Comment (ppekrol) [inline src/Raven.Server/Documents/AttachmentsStorage.cs]: Can we check this earlier? After checking if transaction is null?

File: src/Raven.Server/Documents/CountersStorage.cs [hunk 2]
```diff
@@ -775,12 +772,32 @@ private static SortedSet<string> GetCountersForDocument(BlittableJsonReaderArray
...
+            {
+                ThrowStorageKeyTooBig(documentId, counterName);
+            }
+        }
+
         private static void ThrowCounterNameTooBig(string name)
         {
             throw new ArgumentException(
-                $"Counter name cannot exceed {MaxCounterNameSize} bytes, but counter name was {Encoding.UTF8.GetByteCount(name)} bytes. The invalid counter name is '{name}'.",
+                $"Counter name cannot exceed {DocumentIdWorker.MaxIdSize} bytes, but counter name was {Encoding.UTF8.GetByteCount(name)} bytes. The invalid counter name is '{name}'.",
```
Comment (ppekrol) [inline src/Raven.Server/Documents/CountersStorage.cs]: I think we can get rid of the GetByteCount here? We are using string.Length for that now?

**Comment by karmeli87** [comment]
We also need to consider if those exceptions are thrown during the `Apply` of the `ClusterStateMachine`.
Is it okay to commit?
if it isn't we have a problem, because we cannot recover from this.
If it is fine to commit, a custom exception `KeyTooBigException` would be better, so we can catch it and treat as "Expected".

File: src/Raven.Server/Documents/AttachmentsStorage.cs (hunk 1 above)
Comment (aviv86) [reply src/Raven.Server/Documents/AttachmentsStorage.cs]: done

File: src/Raven.Server/Documents/CountersStorage.cs (hunk 2 above)
Comment (aviv86) [reply src/Raven.Server/Documents/CountersStorage.cs]: done

**Review by grisha-kotler — APPROVED** [review]

**Review by ppekrol — APPROVED** [review]

**Pull request merged**